[SPARK-31009][SQL] Support json_object_keys function

iRakson opened the pull request:
<!--
Thanks for sending a pull request!  Here are some tips for you:
  1. If this is your first time, please read our contributor guidelines: https://spark.apache.org/contributing.html
  2. Ensure you have added or run the appropriate tests for your PR: https://spark.apache.org/developer-tools.html
  3. If the PR is unfinished, add '[WIP]' in your PR title, e.g., '[WIP][SPARK-XXXX] Your PR title ...'.
  4. Be sure to keep the PR description updated to reflect all changes.
  5. Please write your PR title to summarize what this PR proposes.
  6. If possible, provide a concise example to reproduce the issue for a faster review.
  7. If you want to add a new configuration, please read the guideline first for naming configurations in
     'core/src/main/scala/org/apache/spark/internal/config/ConfigEntry.scala'.
-->

### What changes were proposed in this pull request?
A new function `json_object_keys` is proposed in this PR. This function will return all the keys of the outmost json object. It takes Json Object as an argument.

- If invalid json expression is given, `NULL` will be returned.
- If an empty string or json array is given, `NULL` will be returned.
- If valid json object is given, all the keys of the outmost object will be returned as an array.
- For empty json object, empty array is returned.


We can also get JSON object keys using `map_keys+from_json`.  But `json_object_keys` is more efficient.
```
Performance result for json_object = {"a":[1,2,3,4,5], "b":[2,4,5,12333321]}

Intel(R) Core(TM) i7-9750H CPU @ 2.60GHz
JSON functions:                           Best Time(ms)   Avg Time(ms)   Stdev(ms)    Rate(M/s)   Per Row(ns)   Relative
------------------------------------------------------------------------------------------------------------------------
json_object_keys                                  11666          12361         673          0.9        1166.6       1.0X
from_json+map_keys                                15309          15973         701          0.7        1530.9       0.8X

```
<!--
Please clarify what changes you are proposing. The purpose of this section is to outline the changes and how this PR fixes the issue. 
If possible, please consider writing useful notes for better and faster reviews in your PR. See the examples below.
  1. If you refactor some codes with changing classes, showing the class hierarchy will help reviewers.
  2. If you fix some SQL features, you can provide some references of other DBMSes.
  3. If there is design documentation, please add the link.
  4. If there is a discussion in the mailing list, please add the link.
-->


### Why are the changes needed?
This function will help naive users in directly extracting the keys from json string and its fairly intuitive as well. Also its extends the functionality of spark-sql for json strings.

Some of the most popular DBMSs supports this function.
- PostgreSQL
- MySQL
- MariaDB
<!--
Please clarify why the changes are needed. For instance,
  1. If you propose a new API, clarify the use case for a new API.
  2. If you fix a bug, you can clarify why it is a bug.
-->


### Does this PR introduce any user-facing change?
Yes. Now users can extract keys of json objects using this function.
<!--
If yes, please clarify the previous behavior and the change this PR proposes - provide the console output, description and/or an example to show the behavior difference if possible.
If no, write 'No'.
-->


### How was this patch tested?
UTs added.
<!--
If tests were added, say they were added here. Please make sure to add some test cases that check the changes thoroughly including negative and positive cases if possible.
If it was tested in a way different from regular unit tests, please clarify how you tested step by step, ideally copy and paste-able, so that other reviewers can test and check, and descendants can verify in the future.
If tests were not added, please describe why they were not added and/or why it was difficult to add.
-->

iRakson committed:
[SPARK-31009][SQL] Support json_object_keys function

AmplabJenkins commented:
Can one of the admins verify this patch?

AmplabJenkins commented:
Can one of the admins verify this patch?

iRakson commented:
@HyukjinKwon @cloud-fan @maropu @dongjoon-hyun

MaxGekk commented:
@iRakson Could you show the use case of the function.

MaxGekk reviewed (commented):
@iRakson Could you show the use case of the function.

MaxGekk commented:
We can get the same result using existing functions:
```scala
scala> val df = Seq("""{"a":1, "b":2}""").toDF("json")
df: org.apache.spark.sql.DataFrame = [json: string]

scala> df.select(map_keys(from_json($"json", MapType(StringType, StringType)))).show
+-----------------+
|map_keys(entries)|
+-----------------+
|           [a, b]|
+-----------------+
```
What's the benefit of having the function in public API?

iRakson commented:
> What's the benefit of having the function in public API?

- `json_object_keys` gives better performance than `map_keys+from_json`.
- Users are more familiar with `json_object_keys` as it supported by some of popular DBMSs, so they will find it easier to use.

iRakson commented:
@HyukjinKwon @cloud-fan @maropu 
Please take a look at this as well.

dongjoon-hyun commented:
ok to test

dongjoon-hyun commented:
Do you know how much faster it is?
> json_object_keys gives better performance than map_keys+from_json.

SparkQA commented:
**[Test build #119637 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/119637/testReport)** for PR 27836 at commit [`17caff4`](https://github.com/apache/spark/commit/17caff478589b57149b33cd941eb7155b616192b).

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/24367/
Test PASSed.

SparkQA commented:
**[Test build #119637 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/119637/testReport)** for PR 27836 at commit [`17caff4`](https://github.com/apache/spark/commit/17caff478589b57149b33cd941eb7155b616192b).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:
  * `case class JsonObjectKeys(child: Expression) extends UnaryExpression with CodegenFallback `

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/119637/
Test FAILed.

iRakson committed:
fix

SparkQA commented:
**[Test build #119648 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/119648/testReport)** for PR 27836 at commit [`7a74b30`](https://github.com/apache/spark/commit/7a74b303d209b67adfcdb3b902daed5ff8c3793e).

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/24379/
Test PASSed.

SparkQA commented:
**[Test build #119648 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/119648/testReport)** for PR 27836 at commit [`7a74b30`](https://github.com/apache/spark/commit/7a74b303d209b67adfcdb3b902daed5ff8c3793e).
 * This patch **fails due to an unknown error code, -9**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/119648/
Test FAILed.

iRakson commented:
retest this please

HyukjinKwon commented:
retest this please

HyukjinKwon commented:
@MaxGekk does it look good to you?

SparkQA commented:
**[Test build #120032 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/120032/testReport)** for PR 27836 at commit [`7a74b30`](https://github.com/apache/spark/commit/7a74b303d209b67adfcdb3b902daed5ff8c3793e).

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/24750/
Test PASSed.

maropu commented:
> Do you know how much faster it is?
>> json_object_keys gives better performance than map_keys+from_json.

Has the @dongjoon-hyun comment above been already adderssed? If this func has performance benefits, I think its better to show the perf. numbers in the PR description.

iRakson commented:
> > Do you know how much faster it is?
> > > json_object_keys gives better performance than map_keys+from_json.
> 
> Has the @dongjoon-hyun comment above been already adderssed? If this func has performance benefits, I think its better to show the perf. numbers in the PR description.

I have updated the time required for `map_keys+from_json` and `json_object_keys`. I will also try to add benchmark for all the new JSON functions.
@maropu @HyukjinKwon

SparkQA commented:
**[Test build #120032 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/120032/testReport)** for PR 27836 at commit [`7a74b30`](https://github.com/apache/spark/commit/7a74b303d209b67adfcdb3b902daed5ff8c3793e).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/120032/
Test PASSed.

iRakson commented:
> @MaxGekk does it look good to you?

@MaxGekk Kindly Review this.

MaxGekk commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/jsonExpressions.scala:
```diff
@@ -781,3 +783,64 @@ case class SchemaOfJson(
+    var arrayBufferOfKeys = ArrayBuffer.empty[UTF8String]
+    // this handles `NULL` case
+    if (parser.nextToken() == null) {
+      throw new AnalysisException(s"$prettyName expect a JSON object but nothing is provided.")
```
Is the exception thrown at analysis phase? I think `AnalysisException` should be replaced by `RuntimeException`.

MaxGekk commented:
The function works for literals but I got NPE on a column with JSON objects:
```scala
  test("json_object_keys") {
    spark.sql("""select json_object_keys('{"a": 1, "b": 2, "c": 3}')""").show(false)
    val df = Seq(
      """{"a":1}""",
      """{"a":1, "b":2}""",
      """{"a": 1, "b": 2, "c": 3}""").toDF("json")
    df.show(false)
    val dfKeys = df.selectExpr("json_object_keys(json)")
    dfKeys.show(false)
  }
```
```
+------------------------------------------+
|json_object_keys({"a": 1, "b": 2, "c": 3})|
+------------------------------------------+
|[a, b, c]                                 |
+------------------------------------------+

+------------------------+
|json                    |
+------------------------+
|{"a":1}                 |
|{"a":1, "b":2}          |
|{"a": 1, "b": 2, "c": 3}|
+------------------------+

06:28:11.089 ERROR org.apache.spark.executor.Executor: Exception in task 0.0 in stage 3.0 (TID 3)
java.lang.NullPointerException
	at org.apache.spark.sql.catalyst.InternalRow$.$anonfun$getAccessor$16(InternalRow.scala:152)
...
Caused by: java.lang.NullPointerException
	at org.apache.spark.sql.catalyst.InternalRow$.$anonfun$getAccessor$16(InternalRow.scala:152)
	at org.apache.spark.sql.catalyst.InternalRow$.$anonfun$getAccessor$16$adapted(InternalRow.scala:151)
	at org.apache.spark.sql.catalyst.expressions.BoundReference.eval(BoundAttribute.scala:41)
	at org.apache.spark.sql.catalyst.expressions.JsonObjectKeys.json$lzycompute(jsonExpressions.scala:812)
	at org.apache.spark.sql.catalyst.expressions.JsonObjectKeys.json(jsonExpressions.scala:812)
```

iRakson committed:
review comment fix

AmplabJenkins commented:
Build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/120161/
Test FAILed.

iRakson commented:
> The function works for literals but I got NPE on a column with JSON objects:
> 
> ```scala
>   test("json_object_keys") {
>     spark.sql("""select json_object_keys('{"a": 1, "b": 2, "c": 3}')""").show(false)
>     val df = Seq(
>       """{"a":1}""",
>       """{"a":1, "b":2}""",
>       """{"a": 1, "b": 2, "c": 3}""").toDF("json")
>     df.show(false)
>     val dfKeys = df.selectExpr("json_object_keys(json)")
>     dfKeys.show(false)
>   }
> ```
> 
> ```
> +------------------------------------------+
> |json_object_keys({"a": 1, "b": 2, "c": 3})|
> +------------------------------------------+
> |[a, b, c]                                 |
> +------------------------------------------+
> 
> +------------------------+
> |json                    |
> +------------------------+
> |{"a":1}                 |
> |{"a":1, "b":2}          |
> |{"a": 1, "b": 2, "c": 3}|
> +------------------------+
> 
> 06:28:11.089 ERROR org.apache.spark.executor.Executor: Exception in task 0.0 in stage 3.0 (TID 3)
> java.lang.NullPointerException
> 	at org.apache.spark.sql.catalyst.InternalRow$.$anonfun$getAccessor$16(InternalRow.scala:152)
> ...
> Caused by: java.lang.NullPointerException
> 	at org.apache.spark.sql.catalyst.InternalRow$.$anonfun$getAccessor$16(InternalRow.scala:152)
> 	at org.apache.spark.sql.catalyst.InternalRow$.$anonfun$getAccessor$16$adapted(InternalRow.scala:151)
> 	at org.apache.spark.sql.catalyst.expressions.BoundReference.eval(BoundAttribute.scala:41)
> 	at org.apache.spark.sql.catalyst.expressions.JsonObjectKeys.json$lzycompute(jsonExpressions.scala:812)
> 	at org.apache.spark.sql.catalyst.expressions.JsonObjectKeys.json(jsonExpressions.scala:812)
> ```

I fixed this issue and also added a new test case for this particular issue.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/24875/
Test PASSed.

SparkQA commented:
**[Test build #120162 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/120162/testReport)** for PR 27836 at commit [`ce18e41`](https://github.com/apache/spark/commit/ce18e41666b603ad8336e61a24c7e88bb0a7fb99).

SparkQA commented:
**[Test build #120162 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/120162/testReport)** for PR 27836 at commit [`ce18e41`](https://github.com/apache/spark/commit/ce18e41666b603ad8336e61a24c7e88bb0a7fb99).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/120162/
Test PASSed.

maropu commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/jsonExpressions.scala:
```diff
@@ -796,3 +797,63 @@ case class SchemaOfJson(
+
+  override def eval(input: InternalRow): Any = {
+    try {
+      lazy val json = child.eval(input).asInstanceOf[UTF8String]
```
why we need `lazy` here?

maropu commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/jsonExpressions.scala:
```diff
@@ -796,3 +797,63 @@ case class SchemaOfJson(
+  arguments = """
+    Arguments:
+      * json_object - A JSON object is required as argument. `Null` is returned, if an invalid JSON
+          string is given. `Runtime Exception` is thrown, if null string or JSON array is given.
```
How about this?
```
      * json_object - A JSON object. If it is an invalid string, the function returns null.
              If it is a JSON array or null, a runtime exception will be thrown.
```
btw, the error handling is consistent with the other JSON functions?

iRakson replied:
I checked the other json functions. `IllegalArgumentException` is thrown by `from_json` for invalid inputs and it is better than `RuntimeException`.

maropu commented on sql/core/src/test/scala/org/apache/spark/sql/JsonFunctionsSuite.scala:
```diff
@@ -710,4 +710,10 @@ class JsonFunctionsSuite extends QueryTest with SharedSparkSession {
 
+  test("json_object_keys") {
+    val df = Seq("""{"a": 1, "b": 2, "c": 3}""".stripMargin)
+      .toDF("json")
```
nit: you don't need the line break here.

iRakson committed:
review comment fix

SparkQA commented:
**[Test build #120213 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/120213/testReport)** for PR 27836 at commit [`bac0b75`](https://github.com/apache/spark/commit/bac0b75392cfaa4eb2a2e60755d616f76c38c4b9).

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/24926/
Test PASSed.

iRakson committed:
fix

SparkQA commented:
**[Test build #120215 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/120215/testReport)** for PR 27836 at commit [`de7e02b`](https://github.com/apache/spark/commit/de7e02b00c57384e89ddbf8895acd31e35f42103).

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/24928/
Test PASSed.

SparkQA commented:
**[Test build #120213 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/120213/testReport)** for PR 27836 at commit [`bac0b75`](https://github.com/apache/spark/commit/bac0b75392cfaa4eb2a2e60755d616f76c38c4b9).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/120213/
Test FAILed.

SparkQA commented:
**[Test build #120215 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/120215/testReport)** for PR 27836 at commit [`de7e02b`](https://github.com/apache/spark/commit/de7e02b00c57384e89ddbf8895acd31e35f42103).
 * This patch **fails PySpark pip packaging tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/120215/
Test FAILed.

HyukjinKwon commented:
retest this please

SparkQA commented:
**[Test build #120264 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/120264/testReport)** for PR 27836 at commit [`de7e02b`](https://github.com/apache/spark/commit/de7e02b00c57384e89ddbf8895acd31e35f42103).

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/24975/
Test PASSed.

SparkQA commented:
**[Test build #120264 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/120264/testReport)** for PR 27836 at commit [`de7e02b`](https://github.com/apache/spark/commit/de7e02b00c57384e89ddbf8895acd31e35f42103).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/120264/
Test PASSed.

iRakson commented:
gentle ping @HyukjinKwon @maropu @MaxGekk

MaxGekk commented on sql/core/src/test/resources/sql-tests/results/json-functions.sql.out:
```diff
@@ -346,6 +346,96 @@ org.apache.spark.sql.AnalysisException
+-- !query schema
+struct<>
+-- !query output
+java.lang.IllegalArgumentException
```
Why do you throw `IllegalArgumentException` for `'[1, 2, 3]'` but NULL for `'{"key": 45, "random_string"}'`? It looks slightly inconsistent though both input are invalid.

iRakson replied:
For invalid JSON strings NULL is thrown. `'[1, 2, 3]'` is a valid JSON string. But not a JSON object. So invalid argument.

iRakson replied:
Regarding this issue. I am thinking of implementing a function `ISJSON()`. It will return `NULL` for `null`, `1` for valid JSON string and `0` for invalid JSON string.

This function is supported by most of the major DBMSs.

Currently we return `NULL` for invalid JSON.

In this PR as well as in #27759 , `NULL` is returned for `null` JSON strings. So behaviour might be confusing as we are returning `NULL` for invalid JSON strings as well.

MaxGekk reviewed (approved):
LGTM except one comment

MaxGekk commented on sql/core/src/test/scala/org/apache/spark/sql/JsonFunctionsSuite.scala:
```diff
@@ -710,4 +710,9 @@ class JsonFunctionsSuite extends QueryTest with SharedSparkSession {
   }
 
+  test("json_object_keys") {
+    val df = Seq("""{"a": 1, "b": 2, "c": 3}""".stripMargin).toDF("json")
```
Remove `stripMargin`

iRakson committed:
fix

SparkQA commented:
**[Test build #120508 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/120508/testReport)** for PR 27836 at commit [`db0868a`](https://github.com/apache/spark/commit/db0868a71cdfb28547419b26ae9d5447a786c68d).

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/25214/
Test PASSed.

SparkQA commented:
**[Test build #120508 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/120508/testReport)** for PR 27836 at commit [`db0868a`](https://github.com/apache/spark/commit/db0868a71cdfb28547419b26ae9d5447a786c68d).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/120508/
Test PASSed.

HyukjinKwon reviewed (approved):
Will merge in few days if there are no more comments.

maropu reviewed (approved):
I have no more comment and it looks ok.

dongjoon-hyun commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/jsonExpressions.scala:
```diff
@@ -796,3 +797,64 @@ case class SchemaOfJson(
+      }
+    } catch {
+      case _: JsonProcessingException => null
+    }
```
Apparently, the current implementation seems to have the same limitations: `NullPointerException` and `ClassCastException`. Please add more unit test cases with `null` and `Integer` input.

dongjoon-hyun replied:
I checked.
```
scala> sql("select json_object_keys(null)").show
java.lang.NullPointerException

scala> sql("select json_object_keys(1)").show
java.lang.ClassCastException: class java.lang.Integer cannot be cast to class org.apache.spark.unsafe.types.UTF8String (java.lang.Integer is in module java.base of loader 'bootstrap'; org.apache.spark.unsafe.types.UTF8String is in unnamed module of loader 'app')
```

dongjoon-hyun commented on sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/JsonExpressionsSuite.scala:
```diff
@@ -791,4 +791,42 @@ class JsonExpressionsSuite extends SparkFunSuite with ExpressionEvalHelper with
+    val another_complex_json_object = """{"k1": [1, 2, {"key": 5}], "k2": {"key2": [1, 2]}}"""
+    val empty_json_array = """[]"""
+    val invalid_json_object = """{[1,2]}"""
+    val another_invalid_json_object = """{"key": 45, "random_string"}"""
```
Please use the simpler pattern which I commented as an example in your sister PR.

dongjoon-hyun commented on sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/JsonExpressionsSuite.scala:
```diff
@@ -791,4 +791,42 @@ class JsonExpressionsSuite extends SparkFunSuite with ExpressionEvalHelper with
+      checkEvaluation(JsonObjectKeys(Literal(null_object)), null)
+    }.getCause
+
+    assert(exception.isInstanceOf[IllegalArgumentException])
```
ditto. This should be checked at 818.

dongjoon-hyun commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/jsonExpressions.scala:
```diff
@@ -796,3 +797,64 @@ case class SchemaOfJson(
+  arguments = """
+    Arguments:
+      * json_object - A JSON object. If it is an invalid string, the function returns null.
+          If it is a JSON array or null, a runtime exception will be thrown.
```
Is there a reason why you choose a runtime exception for `null`? In SQL world, for `null` input, `null` return is expected instead of a runtime exception.

iRakson replied:
Actually I used `null` here for null string(`''`).

dongjoon-hyun reviewed (changes requested):
Hi, @iRakson . I'll review again after you revise this PR. Thanks.

cloud-fan commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/jsonExpressions.scala:
```diff
@@ -796,3 +797,64 @@ case class SchemaOfJson(
+        ["f1","f2"]
+  """,
+  since = "3.1.0")
+case class JsonObjectKeys(child: Expression) extends UnaryExpression with CodegenFallback {
```
please implement `ExpectsInputTypes` to specify the expected input types, so that the analyzer can do proper type check.

cloud-fan commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/jsonExpressions.scala:
```diff
@@ -796,3 +797,64 @@ case class SchemaOfJson(
+
+  override def eval(input: InternalRow): Any = {
+    try {
+      val json = child.eval(input).asInstanceOf[UTF8String]
```
most SQL functions return null for null input, I think this function should follow.

iRakson committed:
review comment fix

SparkQA commented:
**[Test build #120808 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/120808/testReport)** for PR 27836 at commit [`7cb2e4f`](https://github.com/apache/spark/commit/7cb2e4fb32ea6ff4ca7b2ea7cacfd4ddbc3a18d3).

iRakson commented:
@dongjoon-hyun @cloud-fan 
I have tried to handle all the review comments. Kindly review the changes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/25507/
Test PASSed.

dongjoon-hyun commented:
Thank you for updating, @iRakson .

dongjoon-hyun commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/jsonExpressions.scala:
```diff
@@ -796,3 +797,71 @@ case class SchemaOfJson(
+  arguments = """
+    Arguments:
+      * json_object - A JSON object. If it is an invalid string, the function returns null.
+          If it is a JSON array, an illegal argument exception will be thrown.
```
Could you sync the PR description and this function accordingly? This function description looks too simple and the PR description seems wrong in the part `Analysis Exception`.
```
If invalid json expression is given, NULL will be returned.
If a null string or json array is given, an Analysis Exception is thrown.
If valid json object is given, all the keys of outer object will be returned as array.
For empty json object, empty array is returned.
```

iRakson replied:
Actually PR description is not updated. I will update it. I will update the function description as well.

dongjoon-hyun replied:
Thanks!

dongjoon-hyun commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/jsonExpressions.scala:
```diff
@@ -796,3 +797,71 @@ case class SchemaOfJson(
 }
+
+/**
+ * A function which returns all the keys of outer JSON object.
```
Since JSON can have a nested structure with multiple `inner`s and `outer`s, shall we use `the outmost` instead of `outer`?

dongjoon-hyun commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/jsonExpressions.scala:
```diff
@@ -796,3 +797,71 @@ case class SchemaOfJson(
+
+  private def getJsonKeys(parser: JsonParser, input: InternalRow): Any = {
+    var arrayBufferOfKeys = ArrayBuffer.empty[UTF8String]
+    // this handles `NULL` case
```
I understand that you meant `parser.nextToken() == null`, but shall we update the comment like `This handles empty string case`?

dongjoon-hyun commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/jsonExpressions.scala:
```diff
@@ -796,3 +797,71 @@ case class SchemaOfJson(
+
+  override def eval(input: InternalRow): Any = {
+    val json = child.eval(input).asInstanceOf[UTF8String]
+    if(json == null) {
```
If you don't mind, please add a comment before this line.

dongjoon-hyun commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/jsonExpressions.scala:
```diff
@@ -796,3 +797,71 @@ case class SchemaOfJson(
+    // this handles `NULL` case
+    if (parser.nextToken() == null) {
+      throw new IllegalArgumentException(
+        s"$prettyName expect a JSON object but nothing is provided.")
```
`expect` -> `expects`?

iRakson committed:
review comment fix

SparkQA commented:
**[Test build #120815 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/120815/testReport)** for PR 27836 at commit [`d296556`](https://github.com/apache/spark/commit/d296556378a8d7867bf0d6dfeaac4cf9338325fb).

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/25514/
Test PASSed.

SparkQA commented:
**[Test build #120816 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/120816/testReport)** for PR 27836 at commit [`bd8f9a4`](https://github.com/apache/spark/commit/bd8f9a44d8abe62e0f42e9564e570236aa40d94a).

iRakson commented:
@dongjoon-hyun 
I handled the comments. Please verify.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/25515/
Test PASSed.

SparkQA commented:
**[Test build #120808 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/120808/testReport)** for PR 27836 at commit [`7cb2e4f`](https://github.com/apache/spark/commit/7cb2e4fb32ea6ff4ca7b2ea7cacfd4ddbc3a18d3).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:
  * `case class JsonObjectKeys(child: Expression) extends UnaryExpression with CodegenFallback`

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/120808/
Test PASSed.

SparkQA commented:
**[Test build #120815 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/120815/testReport)** for PR 27836 at commit [`d296556`](https://github.com/apache/spark/commit/d296556378a8d7867bf0d6dfeaac4cf9338325fb).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/120815/
Test PASSed.

SparkQA commented:
**[Test build #120816 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/120816/testReport)** for PR 27836 at commit [`bd8f9a4`](https://github.com/apache/spark/commit/bd8f9a44d8abe62e0f42e9564e570236aa40d94a).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/120816/
Test PASSed.

cloud-fan commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/jsonExpressions.scala:
```diff
@@ -796,3 +797,74 @@ case class SchemaOfJson(
+      * json_object - A JSON object. If a valid JSON object is given, all the keys of the outmost
+          object will be returned as an array. If it is an invalid JSON string, the function
+          returns null. If it is a JSON array or empty string, an illegal argument exception
+          will be thrown.
```
I'm confused. Why the behavior is different between invalid JSON and JSON array/empty string?

iRakson replied:
For invalid JSON we are returning `NULL` in all the other json functions. So for that i kept it like that.

cloud-fan replied:
Then we should return null for JSON array or empty string, as they are also invalid as the input of this function.

iRakson replied:
Ok, i will make the required changes here.
In that case, i guess i need to do the same in #27759 too ?

cloud-fan replied:
yea

iRakson committed:
review comment fix

SparkQA commented:
**[Test build #120874 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/120874/testReport)** for PR 27836 at commit [`e45e61c`](https://github.com/apache/spark/commit/e45e61cff1793b99e823c407d03eb42f165583e6).

iRakson commented:
@cloud-fan I made the changes.
Kindly review the changes. I will also update the PR description accordingly.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/25566/
Test PASSed.

SparkQA commented:
**[Test build #120874 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/120874/testReport)** for PR 27836 at commit [`e45e61c`](https://github.com/apache/spark/commit/e45e61cff1793b99e823c407d03eb42f165583e6).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/120874/
Test PASSed.

dongjoon-hyun commented:
Hi, @iRakson .
Could you resolve the conflicts?

dongjoon-hyun commented on sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/JsonExpressionsSuite.scala:
```diff
@@ -790,4 +790,24 @@ class JsonExpressionsSuite extends SparkFunSuite with ExpressionEvalHelper with
+      ("""{"key": 45, "random_string"}""", null),
+      ("", null),
+      ("[]", null),
+      ("""[{"key": "JSON"}]""", null)
```
Could you sort the test cases a little meaningfully, please?

dongjoon-hyun commented on sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/JsonExpressionsSuite.scala:
```diff
@@ -790,4 +790,24 @@ class JsonExpressionsSuite extends SparkFunSuite with ExpressionEvalHelper with
+      ("", null),
+      ("[]", null),
+      ("""[{"key": "JSON"}]""", null)
+    ).foreach{
```
nit. `).foreach{` -> `).foreach {`

dongjoon-hyun commented on sql/core/src/test/resources/sql-tests/inputs/json-functions.sql:
```diff
@@ -58,5 +58,22 @@ select schema_of_json('{"c1":01, "c2":0.1}', map('allowNumericLeadingZeros', 'tr
+select json_object_keys(200);
+select json_object_keys('');
+select json_object_keys('{"key": 1}');
+select json_object_keys('{}');
```
Shall we switch line 67 and 68?

dongjoon-hyun commented on sql/core/src/test/scala/org/apache/spark/sql/JsonFunctionsSuite.scala:
```diff
@@ -710,4 +710,9 @@ class JsonFunctionsSuite extends QueryTest with SharedSparkSession {
+    val df = Seq("""{"a": 1, "b": 2, "c": 3}""").toDF("json")
+    val dfKeys = df.selectExpr("json_object_keys(json)")
+    checkAnswer(dfKeys, Row(Array("a", "b", "c")))
+  }
```
We can revert this change from this PR since we have a superset already.

dongjoon-hyun commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/jsonExpressions.scala:
```diff
@@ -796,3 +797,67 @@ case class SchemaOfJson(
+ * A function which returns all the keys of the outmost JSON object.
+ */
+@ExpressionDescription(
+  usage = "_FUNC_(json_object) - returns all the keys of the outmost JSON object as an array.",
```
nit. `returns` -> `Returns`?

dongjoon-hyun commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/jsonExpressions.scala:
```diff
@@ -796,3 +797,67 @@ case class SchemaOfJson(
+        parser => getJsonKeys(parser, input)
+      }
+    } catch {
+      case _: JsonProcessingException => null
```
Just a question: There is no need to handle `IOException` in this PR?

iRakson replied:
Yes, `nextToken()` throws `IOException`.  I will update it with other changes.

dongjoon-hyun commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/jsonExpressions.scala:
```diff
@@ -796,3 +797,67 @@ case class SchemaOfJson(
+    // return null if an empty string or any other valid JSON string is encountered
+    if (parser.nextToken() == null || parser.currentToken() != JsonToken.START_OBJECT) {
+      return null
+    }
```
Shall we move this exception handling part (line 849 ~ 851) from this `getJsonKey` into line 839 (outside of this function)?

iRakson committed:
review comment fix

SparkQA commented:
**[Test build #120963 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/120963/testReport)** for PR 27836 at commit [`7eba58c`](https://github.com/apache/spark/commit/7eba58ccbfab988a6852cb9f308a2d083b34849e).

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/25653/
Test PASSed.

iRakson committed:
fix

iRakson commented:
@dongjoon-hyun I resolved the conflicts and also have tried to handle all the review comments. 
Kindly review.

SparkQA commented:
**[Test build #120964 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/120964/testReport)** for PR 27836 at commit [`faa3571`](https://github.com/apache/spark/commit/faa35719770475e35b5970dc4bd703d3e6cea9d4).

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder-K8s/25654/
Test PASSed.

SparkQA commented:
**[Test build #120963 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/120963/testReport)** for PR 27836 at commit [`7eba58c`](https://github.com/apache/spark/commit/7eba58ccbfab988a6852cb9f308a2d083b34849e).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/120963/
Test PASSed.

SparkQA commented:
**[Test build #120964 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/120964/testReport)** for PR 27836 at commit [`faa3571`](https://github.com/apache/spark/commit/faa35719770475e35b5970dc4bd703d3e6cea9d4).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/120964/
Test PASSed.

dongjoon-hyun reviewed (approved):
+1, LGTM. Thank you, @iRakson and all.
Merged to master for Apache Spark 3.1.0.

svmgarg commented:
> We can get the same result using existing functions:
> 
> ```scala
> scala> val df = Seq("""{"a":1, "b":2}""").toDF("json")
> df: org.apache.spark.sql.DataFrame = [json: string]
> 
> scala> df.select(map_keys(from_json($"json", MapType(StringType, StringType)))).show
> +-----------------+
> |map_keys(entries)|
> +-----------------+
> |           [a, b]|
> +-----------------+
> ```
> 
> What's the benefit of having the function in public API?

We don't need to define the schema in Json_object_keys
